### DISCUSSION TITLE
Question about the clustering the Binary data.

### DISCUSSION THREAD

**Discussion by chenmc1996** [facebookresearch/faiss#3573] [General]
From my understanding, training IndexBinaryIVF already needs to cluster the data to obtain some centroids. 
Then why the faiss.Kmeans doesn't support binary data?
Thanks

**Comment by mdouze** [comment]
Binary data is converted to float before clustering. 
Clustering binary vectors directly (ie. the vector-to-centroid assignment step) could be faster, but it is not implemented.

**Comment by chenmc1996** [comment]
Thanks for the reply.
Could you please elaborate a bit little more?
I want to perform clustering on my binary data based on hamming distance, But I think the conversion is not very straightforward, Since the Euclidean distance for, e.g., uint8 is not positive correlated to the Hamming distance, e.g., Hamming distance between (10 vs. 00) is equal to (01 vs. 00). but the former one has bigger Euclidean distance?
Or can you suggest how can I implement the clustering on my binary data?

**Comment by sef43** [comment]
Hi sorry to bump but I would also like to know more about this question. In IndexBinaryIVF I see that there is a binary to float conversion then a indexFlatL2 is used: https://github.com/facebookresearch/faiss/blob/33c0ba5d002a7cd9761513f06ecc9822079d4a2f/faiss/IndexBinaryIVF.cpp#L251-L269

Does this mean that the BinaryIVF index has its clusters created based on the euclidean distance between the binary vectors? and not the hamming distance between them?

**Comment by chasingegg** [comment]
At the index training stage, binary vectors are converted to float vectors and its centroids are float vectors so we use euclidean distance, when training stage is finished, centroids are converted to binary vectors and start the index assigning stage, assigning is based on the hamming distance between binary vectors.
Also hamming distance == euclidean distance on binary vectors.
In my opinion, binary vectors are converted to float vectors and do the clustering based on float vectors could benefit recall? And hamming distance does not make any sense on float vectors, have to use euclidean distance.
